### PR TITLE
implemented create

### PR DESCRIPTION
**Pull request by JamesSwandale**

### CODE REVIEW

**Comment by JamesSwandale** [comment]
There are probably more elegant solutions to this, please give it a review and let me know any areas I can improve.

File: lib/dictionaries/api.js [hunk 1]
```diff
@@ -33,7 +33,21 @@ module.exports = function(opts){
   });
 
   router.put('/:scope/:uuid/dictionaries/:name.json', parameterize, authorize, function(req, res, next) {
+    var filters = {'meta.uuid': req.params.uuid, 'meta.scope': req.params.scope, name: req.params.name};
+    if (!req.body.name){
```
Comment (jesuspc) [inline lib/dictionaries/api.js]: We should move that logic into a parameterizer

File: lib/dictionaries/api.js [hunk 2]
```diff
@@ -33,7 +33,21 @@ module.exports = function(opts){
...
+
+    var handleResponse = function(dictionary) {
+      if(dictionary) {
+        res.json(serialize(dictionary));
+      } else {
+        res.sendStatus(200);
+      }
+    };
 
+    transactions.create(filters, payload).then(handleResponse);
```
Comment (jesuspc) [inline lib/dictionaries/api.js]: It's weird to call a create action with filters. I think we should name it somehow in which the fact that an update or create is performed. Also payload is not a very nice name for the actual content of the dictionary. I would rather call it dictionaryData or simply dictionary.

File: lib/dictionaries/transactions/create.js [hunk 3]
```diff
@@ -1 +1,27 @@
...
+var Q = require('q');
+
+module.exports = function(opts){
+  var dbConnection = opts.dbConnection;
+
+  return function(filters, payload){
+    var deferred = Q.defer();
+
+    dbConnection().then(function(db){
+      db.collection('dictionaries').findOneAndUpdate(filters, payload, {upsert: true, returnOriginal: false} ).then(function(dict){
```
Comment (jesuspc) [inline lib/dictionaries/transactions/create.js]: There's an extra whitespace after the returnOriginal: false} :)

File: lib/dictionaries/transactions/create.js [hunk 4]
```diff
@@ -1 +1,27 @@
...
+        deferred.resolve(dict.value);
+      })
+  	})
```
Comment (jesuspc) [inline lib/dictionaries/transactions/create.js]: And a weird indentation here

File: lib/dictionaries/transactions/create.js [hunk 5]
```diff
@@ -1 +1,27 @@
...
+    
+    return deferred.promise;
+  };;
```
Comment (jesuspc) [inline lib/dictionaries/transactions/create.js]: Double semicolon

File: lib/dictionaries/transactions/create.js [hunk 6]
```diff
@@ -1 +1,27 @@
...
+};
+
+
+/*
```
Comment (jesuspc) [inline lib/dictionaries/transactions/create.js]: Lets remove the comments. Git is supposed to be used to keep track of that kind of stuff

File: test/integration/createDictionary.spec.js [hunk 7]
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
...
     describe('when successful authorization', function(){
       describe('when payload provided', function(){
-        it('returns a 200', function(){
-
+        beforeEach(function(){
+          this.dictName = 'unnamedPayload';
+        })
+
+        it('returns a 200', function(done){
+          that = this
```
Comment (jesuspc) [inline test/integration/createDictionary.spec.js]: That should be declared as var and should have a semicolon

File: test/integration/createDictionary.spec.js [hunk 8]
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
...
+          var expectedPayload = {"name": "unnamedPayload", "unnamedField" : "unnamedValue"};
+
+          this.doRequest(function(req){
+            req.send(that.dictUnnamedPayload)
+            req.end(function(){
+              dbConnection().then(function(db){
+                db.collection('dictionaries').findOne({name: that.dictName}).then(function(dict){
+                    try{
+                      assert.deepEqual(dict.name, expectedPayload.name);
+                      assert.deepEqual(dict.unnamedField, expectedPayload.unnamedField);
```
Comment (jesuspc) [inline test/integration/createDictionary.spec.js]: Again, payload is not a very nice name

File: test/integration/createDictionary.spec.js [hunk 9]
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
...
+              });
+            });
+          });
         });
 
-        it('returns the created element as json', function(){
-
+        it('returns the created element as json', function(done){
+          var expectedBody = { name: "unnamedPayload", unnamedField: "unnamedValue" };
+          that = this;
```
Comment (jesuspc) [inline test/integration/createDictionary.spec.js]: var that

File: test/integration/createDictionary.spec.js [hunk 10]
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
...
+          });
         });
       });
 
-      describe('when no payload provided', function(){
-        it('returns a 200', function(){
-
+      describe('when no payload provided', function(done){
+        it('returns a 200', function(done){
+          that = this
```
Comment (jesuspc) [inline test/integration/createDictionary.spec.js]: var thar, semicolon and indentation

File: test/integration/createDictionary.spec.js [hunk 11]
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
...
+            this.doRequest(function(req){
+              req.expect(200, done);
+            })
```
Comment (jesuspc) [inline test/integration/createDictionary.spec.js]: A semicolon is missing here

File: test/integration/createDictionary.spec.js [hunk 12]
```diff
@@ -4,42 +4,130 @@ var box = require('../box')(),
...
+        it('creates a record with by with the given name', function(done){
+          that = this;
+          var expectedPayload = {"name": "newDict"};
+
+          this.doRequest(function(req){
+            req.end(function(){
+              dbConnection().then(function(db){
+                db.collection('dictionaries').findOne({name: that.dictName}).then(function(dict){
+                    try{
+                      assert.deepEqual(dict.name, expectedPayload.name);
```
Comment (jesuspc) [inline test/integration/createDictionary.spec.js]: Payload again

**Comment by jesuspc** [comment]
Good work again! There are just some minor issues under my point of view. 

Btw, probably we should start unit testing stuff too

**Comment by JamesSwandale** [comment]
Thanks for the comments, I've fixed the issues, I wasn't sure what to do with "move that logic into a parameterizer" currently I've moved it into a function to deal with later.

**Pull request merged**